File: day_8/puzzle_2/solve.rb
# frozen_string_literal: true

require_relative '../../benchmark/benchmark'

Benchmark.run('day_8_puzzle_2') do
  File.open('./day_8/input.txt', 'r') do |file|
    rows = file.read.split("\n").map { |row| row.split('').map(&:to_i) }
    cols = rows[0].map.with_index { |_, col| rows.map { |row| row[col] } }

    max_length = [rows.length, cols.length].max

    def score_direction(tree, following_trees)
      score = 0

      following_trees.each do |following_tree|
        score += 1

        break if tree <= following_tree
      end

      score
    end

    max_score = 0

    max_length.times do |i|
      next if i.zero?
      next if i == max_length - 2

      row = rows[i]

      max_length.times do |j|
        tree = row[j]

        score =
          score_direction(tree, row[j + 1..]) *
          score_direction(tree, row[0...j].reverse) *
          score_direction(tree, cols[j][i + 1..]) *
          score_direction(tree, cols[j][0...i].reverse)

        max_score = [score, max_score].max
      end
    end

    max_score
  end
end

File: day_8/puzzle_2/solve.ts
import fs from "fs";
import { benchmark } from "../../benchmark/benchmark";

benchmark(() => {
  const file = fs.readFileSync("./day_8/input.txt", { encoding: "utf-8" });

  const rows = file.split("\n").map((row) => row.split("").map(Number));
  const cols = rows[0].map((_, col) => rows.map((row) => row[col]));

  const maxLength = Math.max(rows.length, cols.length);

  let maxScore = 0;

  const scoreDirection = (tree: number, followingTrees: number[]) => {
    let score = 0;

    for (let i = 0; i < followingTrees.length; i++) {
      score++;

      if (tree <= followingTrees[i]) {
        break;
      }
    }

    return score;
  };

  for (let i = 1; i < maxLength - 1; i++) {
    const row = rows[i];

    for (let j = 1; j < maxLength - 1; j++) {
      const tree = row[j];

      const score =
        scoreDirection(tree, row.slice(j + 1)) *
        scoreDirection(tree, row.slice(0, j).reverse()) *
        scoreDirection(tree, cols[j].slice(i + 1)) *
        scoreDirection(tree, cols[j].slice(0, i).reverse());

      maxScore = Math.max(maxScore, score);
    }
  }

  return maxScore;
}, "day_8_puzzle_2");
